### PR TITLE
chore: follow-ups for numeric JSON decimal support and documentation

### PR DESCRIPTION
**Pull request by agavra**
### Description 
Minor follow-ups based on my previous pr and @big-andy-coates' comments. Also some documentation.

### Testing done 
`mvn clean install` and looked at the docs after `make html`

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/developer-guide/syntax-reference.rst [hunk 1]
```diff
@@ -306,6 +306,22 @@ defines a struct with three fields, with the supplied name and type.
...
+KSQL supports fields that are numeric data types with fixed precision and scale:
+
+- **Precision** is the maximum total number of decimal digits to be stored, including values to
+  the left and right of the decimal point. The precision must be greater than 1 - there is no
+  default precision.
+- **Scale** is the number of decimal digits to the right of the decimal points. This number must
+  be greater than 0 and less than the value for Precision.
+
+Converting from decimal data type to any floating point type (``DOUBLE``) may cause loss of
+precision. Converting from decimal data type to either ``INTEGER`` or ``BIGINT`` may cause an
```
Comment (stevenpyzhang) [inline docs/developer-guide/syntax-reference.rst]:
```suggestion
precision. Converting from decimal data type to either ``INTEGER`` or ``BIGINT`` may cause a
```

File: docs/developer-guide/serialization.rst [hunk 2]
```diff
@@ -373,6 +373,34 @@ Field Name Case Sensitivity
...
+Decimals with specified precision and scale are serialized as JSON floating point numbers. For
+example:
+
+.. code:: json
+
+        {
+          "value": 1.12345678912345
+        }
+
+Note that this means that any downstream application that reads JSON output from a KSQL application
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
Any downstream application that reads JSON output from a KSQL application
```

File: docs/developer-guide/serialization.rst [hunk 3]
```diff
@@ -373,6 +373,34 @@ Field Name Case Sensitivity
...
+that produces decimals must ensure that their deserializer will correctly maintain the precision of
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
that produces decimals must ensure that its deserializer correctly maintains the precision of
```

File: docs/developer-guide/udf.rst [hunk 4]
```diff
@@ -109,8 +109,9 @@ Dynamic return type
 ~~~~~~~~~~~~~~~~~~~
 
 UDFs support dynamic return types that are resolved at runtime. This is useful if you want to
-implement a UDF with a non-deterministic return type. A UDF which returns ``BigDecimal``,
-for example, may vary the precision and scale of the output based on the input schema.
+implement a UDF with a non-deterministic return type. A UDF which returns ``BigDecimal`` (which maps
```
Comment (JimGalasyn) [inline docs/developer-guide/udf.rst]:
```suggestion
implement a UDF with a non-deterministic return type. For example, a UDF that returns ``BigDecimal`` (which maps
```

File: docs/developer-guide/udf.rst [hunk 5]
```diff
@@ -109,8 +109,9 @@ Dynamic return type
...
+to the SQL ``DECIMAL`` type), for example, may vary the precision and scale of the output based on
```
Comment (JimGalasyn) [inline docs/developer-guide/udf.rst]:
```suggestion
to the SQL ``DECIMAL`` type) may vary the precision and scale of the output based on
```

File: docs/developer-guide/serialization.rst [hunk 6]
```diff
@@ -373,6 +373,34 @@ Field Name Case Sensitivity
...
+Decimals with specified precision and scale are serialized as JSON floating point numbers. For
+example:
+
+.. code:: json
+
+        {
+          "value": 1.12345678912345
+        }
+
+Any downstream application that reads JSON output from a KSQL application that produces decimals
```
Comment (big-andy-coates) [inline docs/developer-guide/serialization.rst]: Personally, I think this paragraph is superfluous. KSQL is output valid / correct JSON.  Calling out that any downstream app should handle this correctly is just a bit: well dur!  It's also calling out Java specific stuff, like Jackson, where the downstream processing could be anything, (especially as KSQL is for none-java users).

File: docs/developer-guide/syntax-reference.rst [hunk 7]
```diff
@@ -306,6 +306,22 @@ defines a struct with three fields, with the supplied name and type.
...
+
+``DECIMAL(Precision, Scale)``
+
+KSQL supports fields that are numeric data types with fixed precision and scale:
+
+- **Precision** is the maximum total number of decimal digits to be stored, including values to
+  the left and right of the decimal point. The precision must be greater than 1 - there is no
+  default precision.
+- **Scale** is the number of decimal digits to the right of the decimal points. This number must
+  be greater than 0 and less than the value for Precision.
```
Comment (big-andy-coates) [inline docs/developer-guide/syntax-reference.rst]:
```suggestion
  be greater than 0 and less than, or equal to, the value for Precision.
```

Right?  e.g. `.123` has scale and precision of `3`.

File: docs/developer-guide/syntax-reference.rst [hunk 8]
```diff
@@ -306,6 +306,22 @@ defines a struct with three fields, with the supplied name and type.
...
+
+Converting from decimal data type to any floating point type (``DOUBLE``) may cause loss of
```
Comment (big-andy-coates) [inline docs/developer-guide/syntax-reference.rst]: It's probably worth explaining _when_ a decimal would be converted to a double, i.e. explaining that that decimals are implicitly converted to doubles, as needed.

File: docs/developer-guide/syntax-reference.rst [hunk 9]
```diff
@@ -306,6 +306,22 @@ defines a struct with three fields, with the supplied name and type.
...
+precision. Converting from decimal data type to either ``INTEGER`` or ``BIGINT`` may cause a
```
Comment (big-andy-coates) [inline docs/developer-guide/syntax-reference.rst]: How would the user convert a decimal to an integer?  Are you talking about a `CAST`? There's no implicit conversion. Worth explaining.

File: docs/developer-guide/udf.rst [hunk 10]
```diff
@@ -109,8 +109,9 @@ Dynamic return type
 ~~~~~~~~~~~~~~~~~~~
 
 UDFs support dynamic return types that are resolved at runtime. This is useful if you want to
-implement a UDF with a non-deterministic return type. A UDF which returns ``BigDecimal``,
-for example, may vary the precision and scale of the output based on the input schema.
+implement a UDF with a non-deterministic return type. For example, A UDF that returns ``BigDecimal``
+(which maps to the SQL ``DECIMAL`` type) may vary the precision and scale of the output based on
+the input schema.
```
Comment (big-andy-coates) [inline docs/developer-guide/udf.rst]: Worth calling out that its also useful for `STRUCT` too?

File: ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/Record.java [hunk 11]
```diff
@@ -61,7 +61,7 @@ public Record(
     this.topic = topic;
     this.key = key;
     this.value = value;
-    this.jsonValue = jsonValue;
+    this.jsonValue = Optional.ofNullable(jsonValue);
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/Record.java]: You can just make the param `Optional<JsonNode>` - Jackson knows how to handle them. Do this on both constructors and you'll simplify the code later on that's doing `someOptional.orElse(null)`.

File: ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/Record.java [hunk 12]
```diff
@@ -37,7 +37,7 @@
   private final Object value;
   private final Optional<Long> timestamp;
   private final WindowData window;
-  private JsonNode jsonValue;
+  private Optional<JsonNode> jsonValue;
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/Record.java]: I take it were making this change as its valid to have a null value in a stream?

File: ksql-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java [hunk 13]
```diff
@@ -327,7 +328,9 @@ private static void compareKeyValueTimestamp(
     final Object actualValue = actual.value();
 
     final Object expectedKey = expected.key();
-    final JsonNode expectedValue = expected.getJsonValue();
+    final JsonNode expectedValue = expected.getJsonValue()
+        .orElseThrow(() -> new KsqlServerException(
+            "could not get expected value from test record: " + expected));
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java]: why would it be empty?

File: docs/developer-guide/syntax-reference.rst (hunk 7 above)
Comment (agavra) [reply docs/developer-guide/syntax-reference.rst]: yup :)

File: docs/developer-guide/syntax-reference.rst (hunk 9 above)
Comment (agavra) [reply docs/developer-guide/syntax-reference.rst]: I will just remove this line altogether - it's probably more misleading than value that it adds.

File: ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/Record.java (hunk 11 above)
Comment (agavra) [reply ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/Record.java]: this isn't the jackson `RecordNode` - this is the resolved node

File: ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/Record.java (hunk 12 above)
Comment (agavra) [reply ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/Record.java]: no - there are parts of the code that create a `Record` from the actual value without a corresponding spec, those cases will not have the original jsonValue

File: ksql-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java (hunk 13 above)
Comment (agavra) [reply ksql-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java]: see above

**Pull request merged**